Feature/chobk 2656 (#2)
meliguicarvalho/sdk-js: docs/bricks/payment.md

# Payment Brick Documentation

The Payment Brick supports two distinct flows based on user authentication status:

## Flow Selection

> The brick automatically determines which flow to use based on the presence of the `supertoken` property in the initialization settings:

Reviewer: Como estamos vendendo como Fast Payments na documentação do DevSite, não faria sentido replicar essa nomenclatura no código e documentação técnica? O termo 'supertoken' era algo mais interno do ambiente de desenvolvimento

Author: Otimo ponto Isma!
Sendo assim, teriamos que mudar o nome da prop no código.
Tem alguma sugestão de nome?

Reviewer: Desculpa a demora!

Sugiro chamar o produto de `Fast Payments` e a propriedade de `fastPaymentsToken` ou `fastPaymentsHash`

- **Guest Flow**: Used when `supertoken` is not present in initialization settings
- **Supertoken Flow**: Used when `supertoken` is present in initialization settings (for authenticated users)

---

## `BricksBuilder`.create(`brick`, `target`, `settings`)

### Parameters

`brick` | _string_, **REQUIRED**

Selected Brick. Possible values are: `payment`.

<br />

`target` | _string_, **REQUIRED**

Id of the container that the brick will be rendered in. Can be any HTML element.

<br />

`settings` | _object_, **REQUIRED**

The `settings` object has properties to initialize and customize the brick being created. The content varies by flow:

| Setting key      | Type     | Description                                                                                                                                                        |              |
| ---------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------ |

> | `initialization` | `object` | Defines the initialization data. Content varies by flow - see [Guest Flow Settings](#guest-flow-settings) or [Supertoken Flow Settings](#supertoken-flow-settings) | **REQUIRED** |

Reviewer: Acho que mudaria a segunda parte das frases para algo assim:

```suggestion
| `initialization` | `object` | Defines the initialization data. Refer to [Guest Flow Settings](#guest-flow-settings) or [Supertoken Flow Settings](#supertoken-flow-settings) for required fields in each flow. | **REQUIRED** |
```

- Edit capabilities for shipping and billing data
- Enhanced customer confidence through order review
- Support for discounts display

**Best for:** High-value transactions, complex orders with multiple items, enhanced customer experience requirements

---

## Flow Selection

The brick automatically determines which flow to use based on the initialization properties:

```js
// Guest Flow - basic configuration
mp.bricks().create('payment', 'container', {
  initialization: {
    amount: 10000,
    // ... other guest flow settings
  },
  // ...
});

// Fast Payments Flow - fastPaymentToken provided
mp.bricks().create('payment', 'container', {
  initialization: {
    fastPaymentToken: '<USER_FAST_PAYMENT_TOKEN>',
    // ... other fast payments flow settings
  },
  // ...
});

// Review Flow - enableReviewStep and review data provided
mp.bricks().create('payment', 'container', {
  initialization: {
    amount: 10000,
    items: {
      /* items data */
    },
    shipping: {
      /* shipping data */
    },
    // ... other initialization data
  },
  customization: {
    enableReviewStep: true,
    // ... other customization options
  },
  // ...
});
```

---

> ## Quick Comparison

Reviewer 2: Deveríamos adicionar aqui que a taxa de aprovação de pagamentos é melhor? Acho que esse é o principal benefício do supertoken, não?

Author: Nao colocaria, porque nao sabemos se isso é verdade de fato

| `MLM (Mexico)` |
| `MPE (Peru)`   |

---

## Supertoken Flow

The Supertoken Flow is designed for authenticated users who have existing payment methods saved in their Mercado Pago account. This flow provides a streamlined payment experience by leveraging the user's authentication token.

### Usage Example

```js
mp.bricks().create('payment', 'paymentBrick_container', {
  initialization: {
    supertoken: '<USER_SUPERTOKEN>',
  },
  customization: {
    visual: {
      hidePaymentButton: false,
      style: {
        theme: 'default',
      },
    },
  },
  callbacks: {
    onReady: () => {
      // handle form ready
    },
    onSubmit: ({ paymentMethod, formData }) => {
      return new Promise((resolve, reject) => {
        fetch('/process_payment', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(formData),
        })
          .then((response) => response.json())
          .then((response) => {
            // get payment result
            resolve();
          })
          .catch((error) => {
            // get payment result error
            reject();
          });
      });

>     },

Reviewer 2: pro submit de supertoken não vamos devolver mais essas duas props. Vamos devolver direto aquele array que vai dentro da propriedade `transactions` do objeto que vai pra API de order

Author: hmmm ok, perfeito! fica no todo entao!

    onError: (error) => {
      // handle error
    },
    onBinChange: (bin) => {
      // handle BIN change for card validation

>     },

Reviewer 2: não tem onBinChange pra supertoken

Author: Pior que tem.
Quando é realizada a seleção de um cartão salvo, esse callback é chamado.
Na linha 336 do arquivo `src/components/organisms/PaymentForm/PaymentOptionSelectionRow/PaymentOptionSelectionRow.svelte` esse callback é chamado.

| `fingerprint`      | `string` | Device fingerprint for fraud prevention and security validation.                                                                     | **OPTIONAL** |

#### Supertoken Callbacks

The Supertoken Flow requires all three core callbacks for proper payment processing.

| Callback key  | Description                                                            |              | Params                              | Returns         |
| ------------- | ---------------------------------------------------------------------- | ------------ | ----------------------------------- | --------------- |
| `onReady`     | It is called when the brick finishes loading                           | **REQUIRED** | `void`                              | `void`          |
| `onError`     | It is called when there is an error in the Brick                       | **REQUIRED** | `BrickError`                        | `void`          |
| `onSubmit`    | It is called when the user clicks on the submit button                 | **REQUIRED** | `PaymentFormData`, `AdditionalData` | `Promise<void>` |

> | `onBinChange` | It is called when the user fills or update card's BIN (first 8 digits) | **OPTIONAL** | `bin`                               | `void`          |

Reviewer 2: remover onbinchange aqui tb

Author: resposta anterior

> **Note**: In the Supertoken Flow, `onSubmit` is **REQUIRED** because authenticated users expect immediate payment processing capabilities.

#### Supertoken Flow Data Types

##### `PaymentFormData`

```ts
{
  selectedPaymentMethod: 'credit_card';
  formData: CardData;
}
```

> // TODO FIX THE SUBMIT DATA

Reviewer 2: Você já deixou um comentário de TODO, mas acredito que estava se referindo do comentário pra baixo. Precisa mudar também essa parte de cima, pois não vamos devolver esse objeto.

| `visual.font`                    | `string`  | Defines the custom font URL. This only applies to the [PCI fields](../fields.md#fields-module). | **OPTIONAL** |
| `visual.style`                   | `Style`   | Defines custom theme and CSS variables. [See Style section](#style)                             | **OPTIONAL** |
| `visual.hidePaymentButton`       | `boolean` | Hides the payment button and disables the `onSubmit` callback.                                  | **OPTIONAL** |
| `paymentMethods`                 | `object`  | Object that allows installment configuration. Contains `maxInstallments`, `minInstallments`     | **OPTIONAL** |
| `paymentMethods.maxInstallments` | `number`  | Maximum number of installments to be offered to the user                                        | **OPTIONAL** |
| `paymentMethods.minInstallments` | `number`  | Minimal number of installments to be offered to the user                                        | **OPTIONAL** |

---

## Shared Components

### Data Types

#### `BrickError`

```ts
{
  type: 'non_critical' | 'critical';
  message: string;
  cause: ErrorCause;
}
```

#### `ErrorCause`

> The following table lists all possible error causes that can occur in the Payment Brick:

Reviewer 2: essa seção vai precisar ficar pendente até terminarmos, porque ainda tem alguns error causes que estão na minha pull request

| ------------------------------------------ | ---------------------------------------------------------- | ----------------- |
| `window_redirect_was_blocked`              | A required window redirect was blocked by the browser      | Guest             |
| `card_token_creation_failed`               | Failed to create a secure token for the card               | Guest, Supertoken |
| `secure_fields_card_token_creation_failed` | Failed to create a secure token using PCI-compliant fields | Guest, Supertoken |

##### Update Errors

| Error Cause                            | Description                                              | Mode  |
| -------------------------------------- | -------------------------------------------------------- | ----- |
| `amount_is_not_number_in_update`       | The amount provided in the update method is not valid    | Guest |
| `payment_review_invalid_update_object` | The update object provided for payment review is invalid | Guest |

#### `AdditionalData`

```ts
{
    'bin': string,
    'lastFourDigits': string,
    'cardholderName': string,
    'paymentTypeId': string,
}

> ```

Reviewer 2: aqui vai ser só bin e lastFourDigits

Author: certo, pelo que entendo o mesmo, ou quase o mesmo, passa com Cartão Salvo normal. Ja que nao tem o nome nesse cenario. Sendo assim, acha que faz sentido mover tudo isso para sessões separadas ou deixamos aqui e colocamos alguns dados como opcional?

Author: (ignorando nesse contexto a discussão sobre a separação de arquivos)

Reviewer 2: hummm bom ponto, não tinha pensado nisso. O ideal seria corrigir de maneira geral, explicando que pra pagamento com cartões salvos vem determinadas props e para outros fluxos vem outras. Mas acho que vale confirmar com o Will se temos tempo pra isso (acredito que não)
